src/core/quadtree.js

```diff
@@ -0,0 +1,462 @@
+    var it = {x:x, y:y, height:height, width:this.root.bounds.width-x};
+    var sorter = function(a, b) {return a.x < b.x ? -1: (a.x>b.x?1:0);};
+    var side1 = [QNode.TOP_LEFT, QNode.BOTTOM_LEFT];
+    var side2 = [QNode.TOP_RIGHT, QNode.BOTTOM_RIGHT];
+    return this.root.retrieve_iterate(it, sorter, side1, side2, {});
+  };
+  QuadTree.prototype.retrieve_xdec = function (x,y,height) {
+    var x0 = this.root.bounds.x;
+    var it = {x:x0, y:y, height:height, width:x-x0};
+    var sorter = function(a, b) {
+      var ax=a.x+a.width, bx=b.x+b.width;
+      return ax < bx ? 1: (ax>bx?-1:0);};
+    var side1 = [QNode.TOP_RIGHT, QNode.BOTTOM_RIGHT];
+    var side2 = [QNode.TOP_LEFT, QNode.BOTTOM_LEFT];
+    return this.root.retrieve_iterate(it, sorter, side1, side2, {});
+  };
+  
+  /**
+   * Iterate through the items with increasing y that intersect the box by: x, 
+   * y, and width.
+   */
+  QuadTree.prototype.retrieve_yinc = function (x,y,width) {
+    var it = {x:x, y:y, width:width, height:this.root.bounds.height-y};
+    var sorter = function(a, b) { return a.y < b.y ? -1: (a.y>b.y?1:0); };
+    var side1 = [QNode.TOP_LEFT,    QNode.TOP_RIGHT];
+    var side2 = [QNode.BOTTOM_LEFT, QNode.BOTTOM_RIGHT];
+    return this.root.retrieve_iterate(it, sorter, side1, side2, {});
+  };
+  /**
+   * Iterate through the items from the down to up as specified by the 
+   * lower left corner of the bounding box given by item: x, y, and width.
+   */
+  QuadTree.prototype.retrieve_ydec = function (x,y,width) {
+    // When decreasing, we're given bottom left corner, convert to top right.
+    var y0     = this.root.bounds.y;  // Calculate the top-left corner
+    var it = {x:x, y:y0, width:width, height:y-y0};
+    var sorter = function(a, b) {
+        var ay=a.y+a.height, by=b.y+b.height;
+        return ay < by ? 1: (ay>by?-1:0);
+    };
+    var side1 = [QNode.BOTTOM_LEFT, QNode.BOTTOM_RIGHT];
+    var side2 = [QNode.TOP_LEFT,    QNode.TOP_RIGHT];
+    return this.root.retrieve_iterate(it, sorter, side1, side2, {});
+  };
+  
+  function QNode(bounds, depth, maxDepth, maxChildren) {
+    this.bounds   = bounds;  //  bounds
+    this.children = [];     // children contained directly in the node
+    this.nodes    = null;   // subnodes
+    this._maxChildren = maxChildren || 4;
+    this._maxDepth = maxDepth || 4;
+    this._depth = depth || 0;
+  }
+  
+  // We consider "top" to be lower y values (screen coords), but if you're 
+  // using a different coord system, everything works. If these constants 
+  // change, the array in subdivide must too.
+  QNode.TOP_LEFT     = 0;
+  QNode.TOP_RIGHT    = 1;
+  QNode.BOTTOM_LEFT  = 2;
+  QNode.BOTTOM_RIGHT = 3;
+  
+  ///////////
+  // Debugging
+  QNode.prototype.print = function () {
+    var tabs = '';
+    for (var d=0; d < this._depth; d++) {
+      tabs += ' ';
+    }
+    if (this.nodes !== null) {
+      var txt={};
+      var total_elements = 0;
+      txt[QNode.TOP_LEFT]     = 'TOP_LEFT';
```

Comment:
Nit: please do not align the equality signs. This also applies to other places in the code.